### ISSUE TITLE
Error 'invalid_grant' in case of steps 'login' -> 'call api' -> 'refresh token'

### ISSUE THREAD

**Issue by jonnydee** [IdentityModel/IdentityModel.OidcClient#352]
- Version: 5.0.0

**Steps to reproduce the bug with unmodified 'ConsoleClientWithBrowser' client**

1. Start client and follow the login procedure.
2. Call the API with 'c' key.
3. Refresh token with 'r' key.

You should notice the reported 'invalid_grant' error.

**Steps to reproduce the bug with modified 'ConsoleClientWithBrowser' client**

I [modified the 'ConsoleClientWithBrowser' example](https://github.com/jonnydee/IdentityModel.OidcClient/tree/simulate-refreshtoken-stealing) such that I can also use it to simulate an attacker stealing the Refresh Token and see the automatic invalidation of all tokens happen in case an old Refresh Token is being used. It also allows to login again.

1. Start the client and login with 'a' key.
3. Call the API with 'b' key.
4. Refresh token with 'c' key.

You should notice the reported 'invalid_grant' error.

**Additional information**

If you do a 'refresh token' as the very first step after the 'login' you can `call api' and 'refresh token' as often as you like in arbitrary order without any issues (no matter if you use the modified client or not).

**Comment by leastprivilege** [comment]
Hi. 

just so I understand - is this a bug in the sample - or the core library?

**Comment by jonnydee** [comment]
I think the bug is in the 'OidcClient' library. I don't see any suspicious code in the sample application.

**UPDATE:** Now that I think of it again the following scenario described maybe is similar but not comparable and won't help you track down the issue. I'm sorry. Just use the 5.0.0 version of the sample console client and try to reproduce the bug as described in the bug report.

I've updated my modified version above. The attacker (who steals tokens) now does not use 'OidcClient' anymore but only its own 'HttpClient' instance to do requests. And the following steps don't reproduce the bug:

1. Start the client and do a user login with 'a' key.
2. Let attacker steal tokens with 'k' key (and creates a new 'HttpClient' instance for calling the API).
3. Let attacker call API with 'l' key (using previously created 'HttpClient' instance and stolen Access Token).
4. Let attacker refresh his tokens with 'm' key (using his 'HttpClient' instance and stolen Refresh Token).

The attacker successfully refreshed his tokens although it is a similar situation: 'login' -> 'call api' -> 'refresh token'. The difference is that the attacker doesn't use an 'OidcClient' and it doesn't do the login which is still done by the user.

**Comment by leastprivilege** [comment]
I am not sure what exactly you are trying to say?!

What do you mean with "attacker steals token". Tokens in OAuth are typically bearer tokens - whoever has them can use them.

**Comment by jonnydee** [comment]
I'm sorry for this confusion. I shouldn't have mentioned this because it hasn't anything to do with the bug itself.. Just use the 5.0.0 `ConsoleClientWithBrowser` and do the steps mentioned in the title of this bug report. I get an `invalid_grant` as response to the 3rd 'refresh token' step.

In case you still would like to know what I meant with "attacker steals token":

Currently, I'm investigating security implications when using "Authentication Code + PKCE" together with [Refresh Token Rotation](https://auth0.com/docs/secure/tokens/refresh-tokens/refresh-token-rotation). According to how I understood it Refresh Token Rotation works as follows:

1. Every time a new Access Token is obtained using the one-time Refresh Token you get a new Access Token + Refresh Token pair.
2. The old Refresh Token you used for obtaining this new pair of tokens is revoked immediately.

Now assume an attacker manages to somehow get a copy of this new pair of tokens. Both the attacker as well as the authorized user can now use the (identical) Access Token to 'call api'. But now let's assume one party refreshes the token pair. This invalidates the old Refresh Token the other party still possesses. As soon as the party with the invalidated Refresh Token now tries to obtain a new token pair the Authorization Server would realize that it already had been used. As a consequence, it would invalidate all tokens, especially also the new ones that were still valid and the user has to login again. This reduces the time an attacker is able to use its stolen tokens which is the actual goal of the Refresh Token Rotation approach.

It is this scenario I wanted to be able to simulate. So I adapted the `ConsoleClientWithBrowser` accordingly. And while I was playing around with the adapted version I stumbled over the bug I reported here (which materialized when I followed the 'login' -> 'call api' -> 'refresh token' procedure). At first, I thought I did something wrong and in order to double-check this I tried to reproduce this bug in the unmodified client. And it turned out it had the same issue.
